design.md

```diff
@@ -0,0 +1,121 @@
+
+### Not in scope
+* Listing jobs
+* Stopping and restarting processes
+* Task resource management (CPU, memory, etc)
+* Automated certificate creation
+
+## Approach
+The current approach for this project is to create something as simple and straightforward as possible. This means that I will not be using any external libraries or dependencies, and will try to leverage the Linux operating system’s existing tooling as much as possible instead of trying to reinvent the wheel. 
+
+This design will cut a lot of corners and leave a lot of ideal functionality out of the initial implementation. I have noted things that should be considered as `Future Improvements` in each section.
+
+## API
+The API will use gRPC to route requests from the client to the server. The API will run locally on the machine that is being managed from the client.
+
+The API will accept requests for running processes as well as requests to return the status and output of a job. More detailed implementation details for the API can be found in the .proto document.
+
+At a high level the API will have the following methods on the localhost using port 9090:
+
+### Run/Stop Methods
+`RunProcess(RunProcessRequest)`
+`StopProcess(StopProcessRequest)`
+
+### Get Information Methods
+`GetProcessStatus(ProcessStatusRequest)`
+`GetProcessOutput(ProcessOutputRequest)`
+
+The `RunProcess` method will return the process identifier for the command issued to the client, which will be used for the `StopProcess`, `GetProcessStatus`, and `GetProcessOutput` methods.
+
+### Future Improvements
+* Run the server in a separate environment, such as kubernetes, which can allow for defining high availability through replicaSets and would move the API out from the domain of the operating system being managed
+* Utilize some kind of data store (relational or not) to track processes instead of relying on files
+* Include resource management of some kind to keep clients from over-taxing the operating system commands are running on
+* Scope processes to users that map back to the clients on the operating system
+
+## Server Authentication
+
+### mTLS
+The service will utilize mTLS to authenticate and secure requests. For this implementation we will manually create a single local CA certificate that will be used to generate a server side certificate and key pair. These certificates and keys will be stored locally for the server.
+
+The server will enforce TLS version 1.3 for maximum security. The server will prefer the TLS13-AES-256-GCM-SHA384 cipher suite to have the best possible encryption standard while maintaining efficient performance. 
+
+### Scoping Requests
+In order to scope requests to each client we will create a user for each client, and that will be authenticated against the user in the certificate sent from the client. The user on the host will match the first part of the email, before the `@`, which will be compared against the email we receive from the client. 
```

Comment:
> will be authenticated against the user in the certificate

Can you elaborate? Where in the certificate is the user stored?

Reply:
This was noted further down in the Client portion, I've moved it up. It will be stored in the `subjAltName` field of the certificate.